server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockIndexInput.java

```diff
@@ -409,4 +416,47 @@ Builder blockSizeShift(int blockSizeShift) {
             return this;
         }
     }
+
+    /**
+     * Simple class to hold the currently open IndexInput backing an instance
+     * of an {@link OnDemandBlockIndexInput}. Lucene may clone one of these
+     * instances, and per the contract[1], the clones will never be closed.
+     * However, closing the instances is critical for our reference counting.
+     * Therefore, we are using the {@link Cleaner} mechanism from the JDK to
+     * close these clones when they become phantom reachable. The clean action
+     * must not hold a reference to the {@link OnDemandBlockIndexInput} itself
+     * (otherwise it would never become phantom reachable!) so we need a wrapper
+     * instance to hold the current underlying IndexInput, while allowing it to
+     * be changed out with different instances as {@link OnDemandBlockIndexInput}
+     * reads through the data.
+     *
+     * [1]: https://github.com/apache/lucene/blob/8340b01c3cc229f33584ce2178b07b8984daa6a9/lucene/core/src/java/org/apache/lucene/store/IndexInput.java#L32-L33
+     */
+    private static class BlockHolder {
+        private IndexInput block;
```

Comment:
Shouldn't `block` be declared `volatile`? it is cleaned in cleaner's thread but will be queried in other threads ...

Reply:
The cleaner won't touch it until the reference is phantom reachable. I guess I was assuming there would not be any concurrency issues, but maybe that's not a valid assumption. There obviously won't be anything else reading or writing when the cleaner action is running. But I don't actually know if a memory barrier is involved so would the cleaner thread be guaranteed to see the latest value of `block`? I can make it volatile to be safe.

Reply:
Trying to come up with any other possible scenario when `close()` is called but not by cleaner, probably should never happen